File: architecture-skills/object-oriented-programming-paradigm/examples/interface-segregation/interface-segregation.py
"""Interface Segregation Principle in Python.

A fat OfficeMachine protocol forces SimplePrinter to implement scan()
and fax() with NotSupported stubs. The fix splits Printer and Scanner
so each client depends only on the methods it actually calls.
"""

from __future__ import annotations

from typing import Protocol


class NotSupportedError(Exception):
    pass


# --- Violation: one fat interface every device must implement ---


class OfficeMachine(Protocol):
    def print_document(self, path: str) -> None: ...
    def scan_document(self, path: str) -> None: ...
    def fax_document(self, path: str) -> None: ...


class SimplePrinter:
    """Only prints — but the fat interface forces stub methods."""

    def print_document(self, path: str) -> None:
        print(f"print {path}")

    def scan_document(self, path: str) -> None:
        raise NotSupportedError("scan")

    def fax_document(self, path: str) -> None:
        raise NotSupportedError("fax")


def run_office_job(machine: OfficeMachine, path: str) -> None:
    machine.print_document(path)
    for action, fn in (
        ("scan", machine.scan_document),
        ("fax", machine.fax_document),
    ):
        try:
            fn(path)
        except NotSupportedError:
            print(f"{action} attempt: NotSupported")


# --- Fix: small client-specific interfaces ---


class Printer(Protocol):
    def print_document(self, path: str) -> None: ...


class Scanner(Protocol):
    def scan_document(self, path: str) -> None: ...


class BasicPrinter:
    def print_document(self, path: str) -> None:
        print(f"print {path}")


class FlatbedScanner:
    def scan_document(self, path: str) -> None:
        print(f"scan {path}")


class MultiFunctionDevice:
    """Implements both interfaces — no stubs required."""

    def print_document(self, path: str) -> None:
        print(f"print {path}")

    def scan_document(self, path: str) -> None:
        print(f"scan {path}")


def print_job(printer: Printer, path: str) -> None:
    printer.print_document(path)


def scan_job(scanner: Scanner, path: str) -> None:
    scanner.scan_document(path)


if __name__ == "__main__":
    print("-- Violation: SimplePrinter forced to stub scan/fax --")
    run_office_job(SimplePrinter(), "document.pdf")

    print("\n-- Fix: clients depend only on what they need --")
    print_job(BasicPrinter(), "document.pdf")
    scan_job(FlatbedScanner(), "page.tif")
    mfd: Printer = MultiFunctionDevice()
    print_job(mfd, "report.pdf")

File: architecture-skills/object-oriented-programming-paradigm/examples/interface-segregation/interface-segregation.cpp
// Interface Segregation Principle in C++.
//
// A fat OfficeMachine interface forces SimplePrinter to stub scan()
// and fax() with std::runtime_error. The fix splits Printer and Scanner
// so each client depends only on the methods it calls.

#include <iostream>
#include <stdexcept>
#include <string>

// --- Violation: one fat interface every device must implement ---

class OfficeMachine {
public:
    virtual ~OfficeMachine() = default;
    virtual void printDocument(const std::string& path) = 0;
    virtual void scanDocument(const std::string& path) = 0;
    virtual void faxDocument(const std::string& path) = 0;
};

class SimplePrinter : public OfficeMachine {
public:
    void printDocument(const std::string& path) override {
        std::cout << "print " << path << "\n";
    }

    void scanDocument(const std::string& path) override {
        (void)path;
        throw std::runtime_error("scan");
    }

    void faxDocument(const std::string& path) override {
        (void)path;
        throw std::runtime_error("fax");
    }
};

void runOfficeJob(OfficeMachine& machine, const std::string& path) {
    machine.printDocument(path);
    try {
        machine.scanDocument(path);
    } catch (const std::runtime_error&) {
        std::cout << "scan attempt: NotSupported\n";
    }
    try {
        machine.faxDocument(path);
    } catch (const std::runtime_error&) {
        std::cout << "fax attempt: NotSupported\n";
    }
}

// --- Fix: small client-specific interfaces ---

class Printer {
public:
    virtual ~Printer() = default;
    virtual void printDocument(const std::string& path) = 0;
};

class Scanner {
public:
    virtual ~Scanner() = default;
    virtual void scanDocument(const std::string& path) = 0;
};

class BasicPrinter : public Printer {
public:
    void printDocument(const std::string& path) override {
        std::cout << "print " << path << "\n";
    }
};

class FlatbedScanner : public Scanner {
public:
    void scanDocument(const std::string& path) override {
        std::cout << "scan " << path << "\n";
    }
};

class MultiFunctionDevice : public Printer, public Scanner {
public:
    void printDocument(const std::string& path) override {
        std::cout << "print " << path << "\n";
    }

    void scanDocument(const std::string& path) override {
        std::cout << "scan " << path << "\n";
    }
};

void printJob(Printer& printer, const std::string& path) {
    printer.printDocument(path);
}

void scanJob(Scanner& scanner, const std::string& path) {
    scanner.scanDocument(path);
}

int main() {
    std::cout << "-- Violation: SimplePrinter forced to stub scan/fax --\n";
    SimplePrinter simple;
    runOfficeJob(simple, "document.pdf");

    std::cout << "\n-- Fix: clients depend only on what they need --\n";
    BasicPrinter basic;
    FlatbedScanner flatbed;
    MultiFunctionDevice mfd;
    printJob(basic, "document.pdf");
    scanJob(flatbed, "page.tif");
    printJob(mfd, "report.pdf");
    return 0;
}
